### PR TITLE
Allow multiple mappings per domain, helper for providing dynamic domains

### PR DESCRIPTION
**Pull request by ziollek**
What has been done:
- extracting logic of json fields to a dedicated class
- change the single mapping to mappings_list field
- adding a helper for providing dynamic domain mappings related to logical cluster
- rendering related dynamic mapping in logical cluster admin view

refer_mapping_by_labels

### CODE REVIEW

File: vaas/vaas/cluster/models.py [hunk 1]
```diff
@@ -14,7 +15,25 @@
...
+
+    def _get_parsed_field(self, field: Optional[set], values: str) -> Set[str]:
+        if field is None:
+            try:
+                field = set(json.loads(values))
+            except:  # noqa
+                field = set()
+        return field
+
+    def _prepare_set_and_json(self, field: Union[List, Set]) -> Tuple[Set, str]:
```
Comment (awmackowiak) [inline vaas/vaas/cluster/models.py:28]: We use python 3.9 and the Union have shorthand form. Just "|".
```python
Union[List, Set] = List | Set
```
Using that shorthand form is recommended by python.

File: vaas/vaas/cluster/models.py [hunk 2]
```diff
@@ -14,7 +15,25 @@
     vcl_template_comment_validator, name_validator, dc_symbol_validator
 
 
-class LogicalCluster(models.Model):
+class AbsModelWithJsonField:
+
+    def _get_parsed_field(self, field: Optional[set], values: str) -> Set[str]:
```
Comment (awmackowiak) [inline vaas/vaas/cluster/models.py:20]: Optional[**S**et] big S i think here is typo.

File: vaas/vaas/cluster/models.py [hunk 3]
```diff
@@ -105,32 +116,75 @@ def validate_unique(self, exclude=None):
...
-    def _prepare_placeholders(self, labels: Set[str]) -> Dict[str, str]:
+    def is_cluster_related_by_labels(self, cluster: LogicalCluster) -> bool:
+        result = False
+        cluster_labels = set(list(cluster.parsed_labels().keys()))
+        # check if all needed placeholders of any mapping are satisfied by cluster labels
+        for required_labels in self.__parse_placeholders().values():
+            result = result or not required_labels.difference(cluster_labels)
+        return result
+
+    def __prepare_placeholders_arguments(self, labels: Set[str]) -> Dict[str, str]:
```
Comment (awmackowiak) [inline vaas/vaas/cluster/models.py:156]: This is duplicated logic with:
```python
def parsed_labels(self) -> Dict[str, str]:
```
could move it to ABSClass?

File: vaas/vaas/cluster/models.py [hunk 4]
```diff
@@ -105,32 +116,75 @@ def validate_unique(self, exclude=None):
...
+        result = {}
+        if self.type == 'dynamic':
+            for mapping in self.mappings:
+                result[mapping] = {name for _, name, _, _ in string.Formatter().parse(str(mapping)) if name}
+        return result
+
     def __str__(self) -> str:
         return f'{self.domain}'
 
+    def _get_parsed_field(self, field, values):
```
Comment (awmackowiak) [inline vaas/vaas/cluster/models.py:173]: This function also is in ABSClass. We can reused.

File: vaas/vaas/cluster/models.py [hunk 5]
```diff
@@ -105,32 +116,75 @@ def validate_unique(self, exclude=None):
...
+        if field is None:
+            try:
+                field = set(json.loads(values))
+            except:  # noqa
+                field = set()
+        return field
+
+    def _prepare_set_and_json(self, field):
```
Comment (awmackowiak) [inline vaas/vaas/cluster/models.py:181]: Same here. But main different is return values.
Maybe better way to return set(), "" like in ABSClass?

File: vaas/vaas/cluster/models.py (hunk 1 above)
Comment (ziollek) [reply vaas/vaas/cluster/models.py:28]: Zrzut ekranu 2023-03-29 o 08 27 19

File: vaas/vaas/cluster/models.py (hunk 3 above)
Comment (ziollek) [reply vaas/vaas/cluster/models.py:156]: No, it was replaced by parsed_labels and it is not used anymore. I will clean that

File: vaas/vaas/cluster/models.py (hunk 4 above)
Comment (ziollek) [reply vaas/vaas/cluster/models.py:173]: thats right, thx!

File: vaas/vaas/cluster/models.py (hunk 5 above)
Comment (ziollek) [reply vaas/vaas/cluster/models.py:181]: thx

**Review by awmackowiak — APPROVED** [review]

**Review by sobczak-m — APPROVED** [review]

**Pull request merged**